### PR TITLE
Add bf16 autocast

### PR DESCRIPTION
**Pull request by tomMcGrath**
# Description

Adds torch autocasting to the SAE train step, improving throughput.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (see W&B dash)
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and unit tests (acceptance tests not currently in use)

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### Performance Check.

If you have implemented a training change, please indicate precisely how performance changes with respect to the following metrics:
- [x] L0
- [x] CE Loss
- [x] MSE Loss
- [ ] Feature Dashboard Interpretability

W&B dashboard showing no performance regression when autocasting to bf16 and moderate speedup/perf-per-time improvement: https://wandb.ai/tmcgrath/autocast%20testing/workspace

More performance could probably be obtained by compiling and autocasting the LLM and maybe by compiling the SAE. Other hyperparameters are also probably better for performance.

### CODE REVIEW

**Comment by tomMcGrath** [comment]
BTW I tried adding fp16 support but training performance degraded pretty badly so it seems more like a footgun and it also makes configs hard to serialise.
